python/lib/core/dmod/core/meta_data.py

```diff
@@ -757,6 +888,73 @@ def factory_init_from_restriction_collections(cls, data_format: DataFormat, **kw
                           discrete_restrictions=None if len(discrete) == 0 else discrete)
 
+    @classmethod
+    def merge_domains(cls, d1: DataDomain, d2: DataDomain) -> DataDomain:
+        """
+        Merge the two domains into a new combined domain.
+
+        Parameters
+        ----------
+        d1: DataDomain
+            The first domain.
+        d2: DataDomain
+            The second domain.
+
+        Returns
+        -------
+        DataDomain
+            The new merged domain.
+
+        Raises
+        ------
+        ValueError
+            If the formats or constraints of the two domains do not permit them to be merged
+
+        Notes
+        -----
+        For any two domains ``d1`` and ``d2`` that can successfully be merged, and the derived domain ``d3`` equal to
+        ``merge_domains(d1, d2)``,  then it should always be true that ``substract_domains(d3, d2) == d1``.
+
+        See Also
+        --------
+        subtract_domains
+        """
+        # TODO: (later) consider exceptions to format rule and perhaps other behavior, like for composites
+        if d1.data_format != d2.data_format:
+            raise ValueError(f"Can't merge {d2.data_format.name} format domain into one of {d1.data_format.name}")
+
+        # New continuous; taken directly from domain 1 if not present or equal in domain 2; otherwise, by extending
+        new_c_restricts: Dict[StandardDatasetIndex, ContinuousRestriction] = {
+            std_ds_idx: (
+                d1_restrict
+                if std_ds_idx not in d2.continuous_restrictions or d1_restrict == d2.continuous_restrictions[std_ds_idx]
+                else d1_restrict.expand(d2.continuous_restrictions[std_ds_idx])
+            )
+            for std_ds_idx, d1_restrict in d1.continuous_restrictions.items()}
+
+        # Any other indices in d2, just move them over
+        for std_ds_idx in (i for i in d2.continuous_restrictions if i not in new_c_restricts):
+            new_c_restricts[std_ds_idx] = d2.continuous_restrictions[std_ds_idx]
+
+        # And now similarly for discrete
+        new_d_restricts: Dict[StandardDatasetIndex, DiscreteRestriction] = {
+            std_ds_idx: (
+                d1_restrict
+                if std_ds_idx not in d2.discrete_restrictions or d1_restrict == d2.discrete_restrictions[std_ds_idx]
+                else d1_restrict.expand(d2.discrete_restrictions[std_ds_idx])
+            )
+            for std_ds_idx, d1_restrict in d1.discrete_restrictions.items()}
+
+        for std_ds_idx in (i for i in d2.discrete_restrictions if i not in new_d_restricts):
+            new_d_restricts[std_ds_idx] = d2.discrete_restrictions[std_ds_idx]
+
+        return DataDomain(data_format=d1.data_format,
+                          continuous_restrictions=new_c_restricts,
+                          discrete_restrictions=new_d_restricts)
+
+    def __eq__(self, other):
+        return isinstance(other, DataDomain) and self.__hash__() == other.__hash__()
```

Comment:
@aaraney, I'm curious if you had something particular in mind when you suggested this.

Reply:
This is purely a style/idiomatic suggestion. They will desugar to the dunder methods like you have it, but generally it is advised to avoid calling dunder methods unless it can't be avoided. For example, you would likely never write `return self.__str__()`, you would instead write `return str(self)`. It doesnt need to be changed, just is the "pythonic" way for better or worse.

Reply:
Eh, fair enough.